### PR TITLE
Add regional model / LES model loss functions to `credit.losses` 

### PR DESCRIPTION
**Pull request by yingkaisha**
TBD

### CODE REVIEW

**Comment by yingkaisha** [comment]
Done

**Comment by jsschreck** [comment]
I liked your extended comments here on how to use the new scripts. Can you add a documentation page that reiterates the same information?

File: applications/rollout_WRF.py [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+    local_rank, world_rank, world_size = get_rank_info(conf["trainer"]["mode"])
+
+    with mp.Pool(num_cpus) as p:
+        if conf["predict"]["mode"] in ["fsdp", "ddp"]:  # multi-gpu inference
+            _ = predict(world_rank, world_size, conf, p=p)
+        else:  # single device inference
+            _ = predict(0, 1, conf, p=p)
+
+    # Ensure all processes are finished
+    p.close()
```
Comment (djgagne) [inline applications/rollout_WRF.py]: p.close() and p.join() should be inside the with block.

File: applications/rollout_LES.py [hunk 2]
```diff
@@ -0,0 +1,518 @@
...
+    local_rank, world_rank, world_size = get_rank_info(conf["trainer"]["mode"])
+
+    with mp.Pool(num_cpus) as p:
+        if conf["predict"]["mode"] in ["fsdp", "ddp"]:  # multi-gpu inference
+            _ = predict(world_rank, world_size, conf, p=p)
+        else:  # single device inference
+            _ = predict(0, 1, conf, p=p)
+
+    # Ensure all processes are finished
+    p.close()
```
Comment (djgagne) [inline applications/rollout_LES.py]: p.close and p.join should be inside the with block.

File: credit/datasets/les_singlestep.py [hunk 3]
```diff
@@ -0,0 +1,512 @@
...
+    generate_datetime,
+    hour_to_nanoseconds,
+    nanoseconds_to_year,
+    find_key_for_number,
+    subset_patch,
+    filter_ds,
+)
+
+
+class LES_Dataset(torch.utils.data.Dataset):
```
Comment (djgagne) [inline credit/datasets/les_singlestep.py]: Python class naming convention uses no underscores between elements of class names. Should be `LESDataset`

File: credit/datasets/wrf_multistep.py [hunk 4]
```diff
@@ -0,0 +1,496 @@
...
+        """
+        Computes the number of batches per epoch for a given batch size.
+
+        Returns:
+        - int: Number of batches per epoch.
+        """
+        return self.num_indices_per_rank
+
+
+class WRF_MultiStep(torch.utils.data.Dataset):
```
Comment (djgagne) [inline credit/datasets/wrf_multistep.py]: Please follow Python conventions with `WRFMultiStep` instead of `WRF_Multistep`.

File: credit/datasets/wrf_singlestep.py [hunk 5]
```diff
@@ -0,0 +1,734 @@
...
+        if self.transform:
+            sample = self.transform(sample)
+
+        # assign sample index
+        sample["index"] = index
+
+        return sample
+
+
+class WRF_Predict(torch.utils.data.IterableDataset):
```
Comment (djgagne) [inline credit/datasets/wrf_singlestep.py]: WRFPredict instead.

File: credit/datasets/wrf_singlestep.py [hunk 6]
```diff
@@ -0,0 +1,734 @@
...
+    find_key_for_number,
+    subset_patch,
+    next_n_hour,
+    previous_hourly_steps,
+    encode_datetime64,
+    filter_ds,
+)
+
+
+class WRF_Dataset(torch.utils.data.Dataset):
```
Comment (djgagne) [inline credit/datasets/wrf_singlestep.py]: WRFDataset instead.

File: credit/models/dscale_wrf.py [hunk 7]
```diff
@@ -0,0 +1,540 @@
...
+            padding_left : pad_lon - padding_right,
+        ]
+
+        # concat
+        x = torch.cat([shortcut, x], dim=1)  # B 2*C Lat Lon
+        x = self.up(x)
+        return x
+
+
+class Dscale_Tansformer(BaseModel):
```
Comment (djgagne) [inline credit/models/dscale_wrf.py]: DscaleTransformer instead.

File: credit/models/swin_wrf.py [hunk 8]
```diff
@@ -0,0 +1,489 @@
...
+            padding_left : pad_lon - padding_right,
+        ]
+
+        # concat
+        x = torch.cat([shortcut, x], dim=1)  # B 2*C Lat Lon
+        x = self.up(x)
+        return x
+
+
+class WRF_Tansformer(BaseModel):
```
Comment (djgagne) [inline credit/models/swin_wrf.py]: WRFTransformer instead.

File: credit/trainers/__init__.py [hunk 9]
```diff
@@ -8,6 +8,10 @@
 from credit.trainers.trainer_downscaling import Trainer as Trainer404
 from credit.trainers.ic_optimization import Trainer as TrainerIC
 
+from credit.trainers.trainerLES import Trainer as TrainerLES
+from credit.trainers.trainerWRF import Trainer as TrainerWRF
+from credit.trainers.trainerWRF_multi import Trainer as TrainerWRF_Multi
```
Comment (djgagne) [inline credit/trainers/__init__.py]: TrainerWRFMulti instead

File: credit/transforms/__init__.py [hunk 10]
```diff
@@ -0,0 +1,81 @@
+import logging
+import numpy as np
+from torchvision import transforms as tforms
+
```
Comment (djgagne) [inline credit/transforms/__init__.py]: We should be applying the same class renaming conventions here too.

**Comment by djgagne** [comment]
All checks pass. Merging.
